src/go/cmd/review.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"slices"
+	"sort"
+	"strings"
+)
+
+var errExternalModule = errors.New("reviewed module exports a type defined in a different repository")
+
+// Review represents an apiview review of an Azure SDK for Go module
+type Review struct {
+	// modules maps module paths to Modules implicated in this API review. It
+	// contains only the reviewed module in most cases, however it contains
+	// more when the reviewed module exports types defined in another module.
+	modules map[string]*Module
+	// name of the APIView review e.g. "sdk/azcore"
+	name string
+	// path on disk to the reviewed module
+	path string
+	// reviewed is the module being reviewed
+	reviewed *Module
+}
+
+// NewReview creates a Review for the module at path p
+func NewReview(p string) (*Review, error) {
+	m, err := NewModule(p)
+	if err != nil {
+		return nil, err
+	}
+	r := &Review{
+		modules: map[string]*Module{},
+		name:    getPackageNameFromModPath(m.ModFile.Module.Mod.Path),
+		path:    p,
+	}
+	fmt.Println("Reviewing", r.name)
+	err = r.AddModule(m)
+	return r, err
+}
+
+// AddModule adds a module to the review. Call this to add a module that exports
+// a type the reviewed module exports by alias.
+func (r *Review) AddModule(m *Module) error {
+	if _, ok := r.modules[m.ModFile.Module.Mod.Path]; ok {
+		return fmt.Errorf("module %s already exists in index", m.ModFile.Module.Mod.Path)
+	}
+	if len(r.modules) == 0 {
+		r.reviewed = m
+	}
+	r.modules[m.ModFile.Module.Mod.Path] = m
+	return nil
+}
+
+func (r *Review) Review() (PackageReview, error) {
+	r.resolveAliases()
+
+	tokenList := &[]Token{}
+	nav := []Navigation{}
+	diagnostics := []Diagnostic{}
+	packageNames := []string{}
+	for name, p := range r.reviewed.Packages {
+		// we use a prefixed path separator so that we can handle the "internal" module.
+		//  internal/dig
+		//  internal/errorinfo
+		//  etc.
+		// for other modules, we skip /internal subdirectories
+		//  azcore/internal/...
+		if strings.Contains(p.relName, "/internal") || p.c.isEmpty() {
+			continue
+		}
+		packageNames = append(packageNames, name)
+	}
+	sort.Strings(packageNames)
+	for _, name := range packageNames {
+		p := r.reviewed.Packages[name]
+		n := p.relName
+		makeToken(nil, nil, "package", TokenTypeMemberName, tokenList)
+		makeToken(nil, nil, " ", TokenTypeWhitespace, tokenList)
+		makeToken(&n, nil, n, TokenTypeTypeName, tokenList)
+		makeToken(nil, nil, "", TokenTypeNewline, tokenList)
+		makeToken(nil, nil, "", TokenTypeNewline, tokenList)
+		// TODO: reordering these calls reorders APIView output and can omit content
```

Comment:
:O

Reply:
😇 it was like this when I got here. I'm moving this code but the problem has existed since the beginning. It could get fixed as part of #8654